### ISSUE TITLE
form fields question

### ISSUE THREAD

**Issue by mwaschkowski** [galkahana/HummusJS#46]
Hi,

Terribly sorry if this is the wrong place, but I'm wondering if this library will allow me to work with form fields and determine the location of them in on a PDF page?

Thank you,

Mark

**Comment by galkahana** [comment]
It is a good place as any other :)

hummusJS has a general parser  that will allow you to read PDF objects.You can read about it here - https://github.com/galkahana/HummusJS/wiki/Parsing.
So yes, you can. it may require some PDF studying.

In your case you are looking for form fields.
Seems like form fields representations on a specific page are defined through "Widget Annotation"
An Annotation contains its position, so what you should do is pretty much this:

- load the PDF to the parser object:
````javascript
var hummus = require('hummus');
var pdfReader = hummus.createReader('./TestMaterials/XObjectContent.PDF');
````
- grab a page dictionary object, say the first and grab its annotations array
````javascript
var pagePDFDictionary = pdfReader.parsePage(0).getDictionary();
if(!pagePDFDictionary.exists('Annots'))
    return;

var annotsArray = pdfPreader.queryDictionaryObject(pagePDFDictionary,'Annots'));
````

- iterate the annotations array looking for "Widget" annotations, and using their Rect entry to realize the position
````javascript
for(var i=0;i<annotsArray.getLength();++i)
{
    var annotationJSDict = pdfPreader.queryArrayObject(annotsArray ,i).toJSObject();
    if(annotationJSDict.Subtype == 'Widget')
   {
      // do something with Rect which will contain the pos and dims
       console.log('annotation bounding box',annotationJSDict.Rect)
   }
   
}
````


Note that this has only to do with the form fields appearance on pages. for their values you should access the global acroform object accessible via the catalog, iterating the fields and getting the required values.

**Comment by mwaschkowski** [comment]
OK, thanks very much! I'm going to give this a try today or tomorrow. If it works it'll be fantastic, been struggling with this.

Cheers,

Mark

**Comment by mwaschkowski** [comment]
Hi,

OK, I gave this a try but ran into some difficulties. This outputted '{}'

  console.log('annotation bounding box',annotationJSDict.Rect);

when I tried this:

  console.log('annotation bounding box', pdfReader.queryArrayObject(rect, 0));

then I got a value, but not sure what value it was.

I tried to iterate the PDFArray object by getting its keys, but that didn't work:

            for (var i = rect.length; i--; ) {
                var key = rect[i];
                console.log("Key=" + key);
            }

so I tried to get the object :

  var temp = pdfReader.queryArrayObject(rect, 0);

but the result is a PDFReal object which I'm not sure what to do with. When I printed it out to the console I got:

  annotation bounding box { value: 515.443 }

Looks like going in the right direction, but how can I get the dimensions?

Thanks so much!

Mark

**Comment by mwaschkowski** [comment]
OK, made a bit of headway with the following:

        if (annotationJSDict.Subtype == 'Widget') {
            // do something with Rect which will contain the pos and dims
            console.log('annotation bounding box',annotationJSDict.Rect);
            var rect = annotationJSDict.Rect; //PDFArray
            var fieldName = annotationJSDict.T.value;
            var x = pdfReader.queryArrayObject(rect, 0);
            var y = pdfReader.queryArrayObject(rect, 1);
            
            console.log('annotation bounding box for field', fieldName);
            console.log('annotation bounding box', x);
            console.log('annotation bounding box', y);
        }

which resulted in:

annotation bounding box for field agencyAddress
annotation bounding box { value: 21.625 }
annotation bounding box { value: 668.063 }
annotation bounding box {}
annotation bounding box for field agencyName
annotation bounding box { value: 21.6 }
annotation bounding box { value: 712.8 }

so I'm guessing there is 4 indexes for a rectangle and we just have to get all of them to get the position:

var y = pdfReader.queryArrayObject(rect, 1); //1-4?

Thanks,

Mar

**Comment by mwaschkowski** [comment]
Assuming I can get dimensions like:

cxt.drawRectangle(left,bottom,width,height,[{options}])

from here:

https://github.com/galkahana/HummusJS/wiki/Show-primitives

?

**Comment by galkahana** [comment]
hi. sorry i didn't get to this earler.
yeah, rect is a PDFArray object.
you can turn it onto a simple js array with toJSArray

so just go

    console.log('annotation bounding box',annotationJSDict.Rect.toJSArray())

and you should get the coordinates array. should give you [left,bottom,right,top].
not that we are talking pdf/ps coordinates (cartesian), so the vector are from the left,bottom of the page, and not like in computers where it's from the left,top.

Cheers,
Gal.

**Comment by mwaschkowski** [comment]
OK, great thanks. After playing around with it for a bit it all worked out. Thanks for your help!

Mark

**Comment by fedyunin** [comment]
Hello! I work on the same,

I need retrieve information about fields, especially field type. I retrieve fields from Annots as in example above.

Above you wrote about:
>Note that this has only to do with the form fields appearance on pages. for their values you should access the global acroform object accessible via the catalog, iterating the fields and getting the required values.

Could you please clarify how to get  acroform object via the catalog. How to get access to catalog?

Thanks!

**Comment by galkahana** [comment]
this should get you to the catalog object:

    var pdfReader = hummus.createReader('myFile.PDF');
    var catalog = pdfReader.queryDictionaryObject(pdfReader.getTrailer(),'Root');


this should get you to the acroform:

    var acroform = pdfReader.queryDictionaryObject(catalog,'AcroForm');

now you can check if exists, to determine if theres a form (if(acroform) {})
from there look at fields through:

var fieldsRoot = pdfReader.queryDictionaryObject(acroform ,'Fields');

etc.

you may need to look into the pdf specs notes on fields to understand how to move on (getting V values, doing a tree iteration etc.). we can either discuss, or you can read in the specs in section 8.6, interactive forms (you can download from here https://app.box.com/s/vz6lewz3eitovlvsshgr.

Regards,
Gal.

**Comment by fedyunin** [comment]
Hi, thank you for answer with details.

There are next questions:

1)
I tried use fieldRoot, but it was unsuccessfull. I can't find correct object which I need (field object related to annotation object).

I tried next approach:
Annotation object has field Parent. I use pdfReader.parseNewObject(objectId) for retrieve parent object: 

 fieldJSDictObject = pdfReader.parseNewObject(annotationJSDict.Parent.getObjectID()).toJSObject();

Is it correct? I found that some Annotation objects doesn't have value for field Parent.

Also in the PDF specs (thank you for link to it) for Annotation there are no Parent field. But there is 
StructParent field. Are they the same?

2)
There is other issue: when I parse PDF receive name of field with incorectly decoded value (maybe wrong charset): e.g: ��. PDF file contains only english text. 
Why this happens? Should I check charset of PDF file or something other?


Thanks!
